src/rules/noUnnecessaryTypeAssertionRule.ts

```diff
@@ -92,9 +94,64 @@ class Walker extends Lint.AbstractWalker<string[]> {
 
         const uncastType = this.checker.getTypeAtLocation(node.expression);
         if (uncastType === castType) {
-            this.addFailureAtNode(node, Rule.FAILURE_STRING, node.kind === ts.SyntaxKind.TypeAssertionExpression
+            // In some cases, this is still not enough; the type in the
+            // assertion can actually impact the type in the expression being
+            // asserted. To avoid these false positives, we create a new
+            // ts.Program with the edited file, to see if the assertion is still
+            // unnecessary.
+            // This can be a bit slow, so we still want to guard it by the more
+            // basic check above.
+            const replacement = node.kind === ts.SyntaxKind.TypeAssertionExpression
                 ? Lint.Replacement.deleteFromTo(node.getStart(), node.expression.getStart())
-                : Lint.Replacement.deleteFromTo(node.expression.getEnd(), node.getEnd()));
+                : Lint.Replacement.deleteFromTo(node.expression.getEnd(), node.getEnd());
+            const sourceFile = this.sourceFile;
+            const modifiedHost = ts.createCompilerHost(this.program.getCompilerOptions());
+            const modifiedSourceFile =
+                ts.createSourceFile(
+                    sourceFile.fileName,
+                    replacement.apply(sourceFile.text),
+                    sourceFile.languageVersion);
+            const oldGetSourceFile = modifiedHost.getSourceFile;
+            modifiedHost.getSourceFile = function(fileName: string, ...args: any[]) {
+                if (fileName === sourceFile.fileName) {
+                    return modifiedSourceFile;
+                }
+                // tslint:disable-next-line:no-unsafe-any Passing args along to original function.
+                return oldGetSourceFile.apply(this, [fileName, ...args]);
```

Comment:
This will be VERY slow even in small projects. See https://github.com/palantir/tslint/pull/2763#issuecomment-333235157. The difference here is, that this implementation reads all but the currently linted sourcefile from disk instead of using the content of the original sourcefile. That makes it even worse.
Also note that #2763 creates a new `Program` for every linted SourceFile. This rule does so for every potential failure.

Reply:
I tried it internally and it took about 0.4 seconds on a medium-large project. I think these cases should be very rare, so this "slow path" will almost always only occur with true positives, which should then (hopefully) get removed.

Reply:
> I think these cases should be very rare

Except for people who just enabled the rule. They will probably think the process ran into an endless loop.

In addition there are projects that disable specific rules for certain (often legacy) files via `tslint:disable` comments. That doesn't turn off the rule. It just mutes its output.